enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,642 @@
+should include `GO_COMPLIANCE_POLICY="exempt_all"` in the environment when
+compiling an extension binary. This will inhibit the wrapper from 
+modifying `CGO_ENABLED` or other compiler arguments required for static compilation.
+
+#### Binary Extraction
+After discovering a test extension binary, the origin test framework will extract the binary
+from the container image which carries it and store it in /tmp storage of the pod in which
+it is running.
+
+If the binary-path ends in `.gz`, the binary will be decompressed.
+
+#### Extension Interface
+Test Extension binaries must implement a well-defined interface through which information about the
+tests the binary offers can be discovered and executed. This interface includes several command line
+verbs the binary must implement, arguments those verbs must support, and output formats the binary
+must adhere to.
+
+Running an extension binary will output the following help text for the initial version of the interface. 
+```
+info      - Output test contribution extension version and metadata.
+list      - Output tests supported by this extension.
+run-test  - Run one or more tests and output results.
+config    - Component configuration management.
+```
+
+##### Info - Extension Metadata
+The extension interface will evolve over time. For this reason, the `info` verb
+will output information about which version of the interface the binary supports. The origin
+framework will support some level of skew in the interface versions and invoke the binaries
+with verbs & arguments consistent with the interface version they support.
+
+Info also provides information to origin about how to construct new or contribute new
+test suites.
+
+Annotated example `info` output is provided below.
+```python
+{
+    # The extension interface version the binary implements.
+    # The details of the interface and version will usually be
+    # unnecessary for a contributor to understand as they will
+    # vendor in the majority of the implementation from the
+    # "Test Extension Support Module" explained later.
+    "apiVersion": "1",
+
+    "source": {
+        # The commit from which this binary from compiled.
+        "commit": "87fdbaa"
+    },
+
+    # A single extension can carry tests for multiple
+    # components. However, each invocation of the binary
+    # can only represent and act on information relative
+    # to a single component.
+    # If a binary carries tests for multiple components,
+    # it must register itself multiple times, with
+    # different pre-verb arguments which openshift-tests
+    # will provide verbatim with every invocation; e.g.
+    # if the binary registers with "--component hyperkube",
+    # verb invocations will be as follows:
+    #    extension-binary --component hyperkube info
+    #    extension-binary --component hyperkube list ...
+    "component": {
+        "product": "openshift",
+        "type": "payload",
+        "name": "hyperkube"
+    },

```

Comment:
What's the value of an additional grouping of tests beyond suite? 

What if tests are cross-component? I guess we still ask there's a primary owner, I'm just wondering why we need the additional slicing of tests

Reply:
> Does this get exposed in the test name at all?
No. The test name is the simple, human readable string. Test ID, which openshift/origin derives from test metadata, is something like  `<product>-<component-type>-<component-name>-<original test name>" and should be unique across extensions and consistent across time (if the test name changes).

> What's the value of an additional grouping of tests beyond suite?
The intent is for non-OpenShift products to participate in the testing.  Component Readiness may provide views for different products.